### PR TITLE
Fix solver competition reading

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Fixes a bug where transaction hashes from different environment (prod/staging) are shown on solver competition endpoint.

Example: https://api.cow.fi/mainnet/api/v1/solver_competition/9861815

Here, transaction `0xb608f1e2a644befda65a6a87ce6b93fc8a3441d9729b8ce4b2a2db02c1247e60` is from production enviroment and valid one, while `0xc85cd0ad8e3aac389836a71da9795bdd1bedce7b21da1950b48f93165c9f6be8` is from staging environment and mistakenly added to result.

The reason why this happens is:
1. The solver competition with the same `auction_id` can be saved in both environments, regardless if the competition ended up with mined settlement (which is fine).
2. `settlement::Observer` saves the relationship <auction, settlement> even for settlements from another environment (this is a decision made long time ago, before refactoring, [reference code here](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/infra/persistence/mod.rs#L654-L660) )

But, what this PR introduces is leveraging the fact that, for settlement from another environment, there is no matching entry in `settlement_observations` table, [reference code here](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/infra/persistence/mod.rs#L682-L693). Note how the saving is executed only if the `settlement` exists, which if [false](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/domain/settlement/observer.rs#L90-L92) for settlements from another environment.

# Changes

- [ ] Update reading functions for solver competition, to exclude transaction hashes for settlements from another environment.

## How to test
Updated unit test.
Existing e2e tests.

Manually tested all updated functions on production database, works as expected.

### CODE REVIEW

File: crates/database/src/solver_competition.rs [hunk 1]
```diff
@@ -38,10 +38,14 @@ pub async fn load_by_id(
     id: AuctionId,
 ) -> Result<Option<LoadCompetition>, sqlx::Error> {
     const QUERY: &str = r#"
-SELECT sc.json, sc.id, COALESCE(ARRAY_AGG(s.tx_hash) FILTER (WHERE s.tx_hash IS NOT NULL), '{}') AS tx_hashes
+SELECT sc.json, sc.id, COALESCE(ARRAY_AGG(s.tx_hash) FILTER (WHERE so.block_number IS NOT NULL), '{}') AS tx_hashes
 FROM solver_competitions sc
 -- outer joins because the data might not have been indexed yet
 LEFT OUTER JOIN settlements s ON sc.id = s.auction_id
+-- exclude settlements from another environment for which observation is guaranteed to not exist
+LEFT OUTER JOIN settlement_observations so
```
Comment (sunce86) [inline crates/database/src/solver_competition.rs:46]: `LEFT OUTER JOIN` makes sure there is no change in logic whether solver competition entry is returned or not

File: crates/database/src/solver_competition.rs [hunk 2]
```diff
@@ -38,10 +38,14 @@ pub async fn load_by_id(
     id: AuctionId,
 ) -> Result<Option<LoadCompetition>, sqlx::Error> {
     const QUERY: &str = r#"
-SELECT sc.json, sc.id, COALESCE(ARRAY_AGG(s.tx_hash) FILTER (WHERE s.tx_hash IS NOT NULL), '{}') AS tx_hashes
+SELECT sc.json, sc.id, COALESCE(ARRAY_AGG(s.tx_hash) FILTER (WHERE so.block_number IS NOT NULL), '{}') AS tx_hashes
```
Comment (sunce86) [inline crates/database/src/solver_competition.rs:41]: Return transaction hash only if `settlement_observation` entry exists.

**Review by MartinquaXD — APPROVED** [review]
Nice, a fix without having to update the db at all. 🚀 

Also thanks for adding a test. 👍

**Comment by sunce86** [comment]
The only drawback of this solution is that `settlement_observation` table exists from 1 March of 2023, so for competitions ran before that date, this logic will not work 100% (transaction hashes will be missing).

But I think this is a trade off we can live with.

**Comment by MartinquaXD** [comment]
I agree that this drawback shouldn't block this fix.

**Review by squadgazzz — APPROVED** [review]
> The only drawback of this solution is that settlement_observation table exists from 1 March of 2023, so for competitions ran before that date, this logic will not work 100% (transaction hashes will be missing).

That’s a very insightful remark.

**Pull request merged**